File: solutions/ocaml/10.ml

let diffs nums = 
  List.map2 (-) nums 
  (0 :: List.(rev % tl % rev) nums)
  |> List.fold_left (fun (d1, d3) -> function
    | 1 -> (d1 + 1, d3)
    | 3 -> (d1, d3 + 1)
    | _ -> (d1, d3)
    ) (0, 0)
  |> fun (d1, d3) -> d1 * (d3 + 1)

let count =
  let rec go acc =
    function
    | [] -> List.hd acc |> snd
    | n :: nums -> 
        go 
        ( (n, List.fold_while (fun sum (m, c) ->
            if m < n - 3 
            then (sum, `Stop)
            else (c + sum, `Continue) 
            ) 0 acc
          ) :: acc
        ) nums
  in go [(0, 1)]

let () = 
  let nums = 
    IO.read_lines_l stdin
  |> List.map Int.of_string_exn
  |> List.sort Int.compare
  in
  Printf.printf "%d %d\n"
  (diffs nums)
  (count nums)

File: solutions/haskell/10.hs
import Data.Functor
import Data.Function
import Data.List
import Debug.Trace

nums :: IO [Int]
nums = getContents <&> lines <&> map read <&> sort

diffs :: [Int] -> Int
diffs nums =
  zipWith (-) nums (0:nums) & foldl (\(d1, d2, d3) d -> 
    case d of
      1 -> (d1+1, d2, d3)
      2 -> (d2, d1+1, d3)
      3 -> (d1, d2, d3+1)
  ) (0, 0, 0) & \(d1, _, d3) -> d1 * (d3 + 1)

count :: [Int] -> Int
count = head . go . (0:) where
  go [_] = [1]
  go (n:l) =
    let cts = go l in
    ( zip l cts & foldr (\(m, ct) sum -> 
        if m > n + 3 then 0 else ct + sum
      ) 0
    ) : cts

-- tail recursive, because I felt like it
count' :: [Int] -> Int
count' = go [(0, 1)] where
    go cts [] = head cts & snd
    go cts (n:l') = go 
      ( (n, foldr (\(m, ct) sum -> 
            if m < n - 3 then 0 else ct + sum
          ) 0 cts
        ) : cts
      ) l'

main :: IO ()
main = do 
  ns <- nums 
  print $ diffs ns
  print $ count ns
  print $ count' ns



